src/crimson/osd/osd.cc

```diff
@@ -391,12 +395,70 @@ seastar::future<> OSD::_send_alive()
   }
 }
 
+/*
+  The OSD's Admin Socket object created here has three servers (i.e. - blocks of commands
+  to handle) registered to it:
+  - OSD's specific commands are handled by the OSD object;
+  - CephContext general commands are registered by the OSD's CephContext, and
+  - there are some common commands registered to be directly handled by the AdminSocket object
+    itself.
+
+  A note regarding the CephContext pointer passed to AdminSocket at creation: that one is used
+  by AdminSocket internally (mostly related to error reporting by legacy code), and is unrelated to
+  the role played by the (same) CephContext object as a command handler.
+*/
+seastar::future<> OSD::start_asok_admin()
+{
+  asok = seastar::make_lw_shared<crimson::admin::AdminSocket>();
+  auto asok_path = local_conf().get_val<std::string>("admin_socket");
+
+  return asok->start(asok_path).then([this] {
+    // register OSD-specific admin-socket hooks
+    admin = std::make_unique<crimson::admin::OsdAdmin>(this);
+    return admin->register_admin_commands();
+  });
+}
+
+/*
+  Note: stop_asok_admin() is executed as part of a destruction process,
+  and may occur even before 'start' is complete. We thus take nothing for granted,
+  and all interfaces are checked for existence.
+*/
+seastar::future<> OSD::stop_asok_admin()
+{
+  return ([this] {
+    if (admin) {
+      return admin->unregister_admin_commands();
+    } else {
+      return seastar::make_ready_future<>();
+    }
+  })().then([this] {
+    if (asok) {
+      return asok->stop().then([this]() {
+        asok.release();
+        return seastar::make_ready_future<>();
+      });
+    } else {
+      return seastar::make_ready_future<>();
+    }
+  }).handle_exception([](auto ep) {
+    logger().error("exception on admin-stop: {}", ep);
+    return seastar::make_ready_future<>();
+  }).finally([] {
+    logger().info("OSD::stop_asok_admin(): Admin-sock service destructed");
+  });
+}
+
 seastar::future<> OSD::stop()
 {
   logger().info("stop");
   // see also OSD::shutdown()
   state.set_stopping();
-  return gate.close().then([this] {
+
+  return gate.close()
+  .then([this] {
```

Comment:
please move `.then([this] {` to the previous line.

Reply:
ok